pieces/simple_csv/rust/src/main.rs

```diff
@@ -0,0 +1,121 @@
+use serde::{Deserialize, Serialize};
+use std::fs;
+use std::path::Path;
+
+#[derive(Debug, Deserialize)]
+#[allow(dead_code)]
+struct CityTemps {
+    city: String,
+    country: String,
+    avg_temp_jan_f: f32,
+}
+
+#[derive(Debug, Serialize)]
+#[allow(dead_code)]
+struct CityTempsModified {
+    city: String,
+    country: String,
+    avg_temp_jan_f: f32,
+    avg_temp_jan_c: f32,
+}
+
+// impl block to convert temp from F to C
+impl CityTempsModified {
+    fn get_celsius(city_temps: &CityTemps) -> Self {
+        let city = &city_temps.city;
+        let country = &city_temps.country;
+        let avg_temp_jan_f = city_temps.avg_temp_jan_f;
+        CityTempsModified {
+            city: city.to_string(),
+            country: country.to_string(),
+            avg_temp_jan_f,
+            avg_temp_jan_c: convert_f_to_c(avg_temp_jan_f),
+        }
+    }
+}
+
+fn convert_f_to_c(temp_f: f32) -> f32 {
+    // Convert from Fahrenheit to Celsius to 1 decimal precision
+    let temp_c = (temp_f - 32.0) * 5.0 / 9.0;
+    (temp_c * 10.0).round() / 10.0
+}
+
+fn read_csv(input_path: &Path) -> Result<Vec<CityTemps>, Box<dyn std::error::Error>> {
+    let contents = fs::read_to_string(input_path).expect("Unable to read from CSV");
+    let mut reader = csv::Reader::from_reader(contents.as_bytes());
+    let data: Vec<CityTemps> = reader.deserialize().collect::<Result<_, _>>()?;
+    println!("Read {} records from {}", data.len(), input_path.display());
+    Ok(data)
+}
+
+fn construct_city_temps_obj(city_temps: Vec<CityTemps>) -> Vec<CityTempsModified> {
+    let mut city_temps_modified: Vec<CityTempsModified> = Vec::new();
+    for item in city_temps.iter() {
+        city_temps_modified.push(CityTempsModified::get_celsius(item));
+    }
+    city_temps_modified
+}
+
+fn write_csv(city_temps_modified: Vec<CityTempsModified>, output_path: &Path) {
+    let mut wtr = csv::Writer::from_path(output_path).expect("Unable to create CSV writer");
+    for item in city_temps_modified.iter() {
+        wtr.serialize(item).expect("Unable to write CSV record");
+    }
+    wtr.flush().expect("Unable to flush CSV writer");
+    println!(
+        "Wrote {} records to {}",
+        city_temps_modified.len(),
+        output_path.display()
+    );
+}
+
+fn main() {
+    let input_path = Path::new("../data/city_temps.csv");
+    let city_temps = read_csv(input_path).expect("Unable to read/open CSV");
+    let city_temps_modified = construct_city_temps_obj(city_temps);
+    // Write the data to a new CSV file
+    let output_path = Path::new("../data/city_temps_modified.csv");
+    write_csv(city_temps_modified, output_path);
+}
+
+#[cfg(test)]
+
+mod tests {
+    use super::*;
+    use approx::assert_abs_diff_eq;
```

Comment:
The `approx` crate is only required during testing. Just like in Python, is there a way in Rust to only call a crate during testing? Not a big deal, but just curious.

Reply:
Yes, you can do `cargo add --dev approx` This will add it to a dev section of the Cargo.toml

```toml
[dev-dependencies]
approx = "x.x.x"
```

Reply:
Great!